vpr/src/pack/greedy_candidate_selector.cpp

```diff
@@ -1245,21 +1264,34 @@ PackMoleculeId GreedyCandidateSelector::get_unrelated_candidate_for_cluster_appa
     }
 
     // Create a queue of locations to search and a map of visited grid locations.
-    std::queue<t_flat_pl_loc> search_queue;
+    std::queue<t_physical_tile_loc> search_queue;
     vtr::NdMatrix<bool, 2> visited({appack_unrelated_clustering_data_.dim_size(0),
                                     appack_unrelated_clustering_data_.dim_size(1)},
                                    false);
     // Push the position of the cluster to the queue.
-    search_queue.push(cluster_gain_stats.flat_cluster_position);
+    t_physical_tile_loc cluster_tile_loc(cluster_gain_stats.flat_cluster_position.x,
+                                         cluster_gain_stats.flat_cluster_position.y,
+                                         cluster_gain_stats.flat_cluster_position.layer);
+    search_queue.push(cluster_tile_loc);
+
+    // Get the max unrelated tile distance for the block type of this cluster.
+    t_logical_block_type_ptr cluster_type = cluster_legalizer.get_cluster_type(cluster_id);
+    float max_dist = appack_ctx_.appack_options.max_unrelated_tile_distance[cluster_type->index];
+
+    // Keep track of the closest compatible molecule and its distance.
+    float best_distance = std::numeric_limits<float>::max();
+    PackMoleculeId closest_compatible_molecule = PackMoleculeId::INVALID();
 
     while (!search_queue.empty()) {
```

Comment:
Could get slow if max_dist is large, as you check everything in this window to see if it is the closest in flat placement space.  Probably there is a way to early exit, since you are expanding out from the cluster_tile_loc (maybe if you are more than 1 unit greater than dist there is no longer any hope and you can break?)